Services/Devices.cs

```diff
@@ -174,15 +210,86 @@ public IDeviceClient GetClient(Device device, IoTHubProtocol protocol)
             await this.GetRegistry().UpdateTwinAsync(deviceId, twin, "*");
         }
 
-        /// <summary> 
-        /// Get IoTHub connection string from either the user provided value or the configuration 
+        /// <summary>
+        /// Create a list of devices
         /// </summary>
-        public void SetCurrentIotHub()
+        public async Task CreateListAsync(IEnumerable<string> deviceIds)
         {
-            string connString = this.connectionStringManager.GetIotHubConnectionString();
-            this.registry = RegistryManager.CreateFromConnectionString(connString);
-            this.ioTHubHostName = IotHubConnectionStringBuilder.Create(connString).HostName;
-            this.log.Info("Selected active IoT Hub for devices", () => new { this.ioTHubHostName });
+            this.SetupHub();
+
+            var batches = this.SplitArray(deviceIds.ToList(), REGISTRY_MAX_BATCH_SIZE).ToArray();
+
+            this.log.Info("Creating devices",
+                () => new { Count = deviceIds.Count(), Batches = batches.Length, REGISTRY_MAX_BATCH_SIZE });
+
+            for (var batchNumber = 0; batchNumber < batches.Length; batchNumber++)
+            {
+                var batch = batches[batchNumber];
+
+                this.log.Info("Creating devices batch",
+                    () => new { batchNumber, batchSize = batch.Count() });
+
+                BulkRegistryOperationResult result = await this.registry.AddDevices2Async(
+                    batch.Select(id => new Azure.Devices.Device(id)), CancellationToken.None);
+
+                this.log.Info("Devices batch created",
```

Comment:
nit: log the batch id so you can better track what failed?

Reply:
👍 also on line 267